core/src/main/java/io/questdb/network/IODispatcherLinux.java

```diff
@@ -45,66 +53,221 @@ public void close() {
+            final long opId = nextOpId();
+            final int fd = context.getFd();
+            int operation = requestedOperation;
+            LOG.debug().$("processing registration [fd=").$(fd)
+                    .$(", op=").$(operation)
+                    .$(", id=").$(opId).I$();
+
+            final SuspendEvent suspendEvent = context.getSuspendEvent();
+            if (suspendEvent != null) {
+                // if the operation was suspended, we request a read to be able to detect a client disconnect 
+                operation = IOOperation.READ;
+            }
+
+            int opRow = pending.addRow();
+            pending.set(opRow, OPM_TIMESTAMP, timestamp);
+            pending.set(opRow, OPM_FD, fd);
+            pending.set(opRow, OPM_ID, opId);
+            pending.set(opRow, OPM_OPERATION, requestedOperation);
+            pending.set(opRow, context);
+
             // we re-arm epoll globally, in that even when we disconnect
             // because we have to remove FD from epoll
-            LOG.debug().$("registered [fd=").$(fd).$(", op=").$(operation).$(", id=").$(id).$(']').$();
-            epoll.setOffset(offset);
-            if (epoll.control(fd, id, EpollAccessor.EPOLL_CTL_MOD, operation == IOOperation.READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT) < 0) {
-                System.out.println("oops2: " + nf.errno());
+            final int epollOp = operation == IOOperation.READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT;
+            if (epoll.control(fd, opId, EpollAccessor.EPOLL_CTL_MOD, epollOp) < 0) {
+                LOG.error().$("epoll_ctl modify operation failure [id=").$(opId)
+                        .$(", err=").$(nf.errno()).I$();
             }
-            offset += EpollAccessor.SIZEOF_EVENT;
 
-            int r = pending.addRow();
-            pending.set(r, M_TIMESTAMP, timestamp);
-            pending.set(r, M_FD, fd);
-            pending.set(r, M_ID, id);
-            pending.set(r, context);
+            if (suspendEvent != null) {
+                // ok, the operation was suspended, so we need to track the suspend event
+                final long eventId = nextEventId();
+                LOG.debug().$("registering suspend event [fd=").$(fd)
+                        .$(", op=").$(operation)
+                        .$(", eventId=").$(eventId)
+                        .$(", suspendedOpId=").$(opId)
+                        .$(", deadline=").$(suspendEvent.getDeadline()).I$();
+
+                int eventRow = pendingEvents.addRow();
+                pendingEvents.set(eventRow, EVM_ID, eventId);
+                pendingEvents.set(eventRow, EVM_OPERATION_ID, opId);
+                pendingEvents.set(eventRow, EVM_DEADLINE, suspendEvent.getDeadline());
+
+                if (epoll.control(suspendEvent.getFd(), eventId, EpollAccessor.EPOLL_CTL_ADD, EpollAccessor.EPOLLIN) < 0) {
+                    LOG.error().$("epoll_ctl add suspend event failure [id=").$(eventId)
```

Comment:
should be critical

Reply:
Changed in bbec37baf832db9a846d04e59ab124f8bb84eadc